pkg/broker/handler/handler_test.go

```diff
@@ -156,6 +160,64 @@ func TestHandler(t *testing.T) {
 	})
 }
 
+type BenchProcessor struct {
+	processors.BaseProcessor
+
+	processed *semaphore.Weighted
+}
+
+func (p *BenchProcessor) Process(ctx context.Context, e *event.Event) error {
+	p.processed.Release(1)
+	return nil
+}
+
+func BenchmarkHandlerReceive(b *testing.B) {
+	for _, eventSize := range []int{0, 1000, 1000000} {
+		b.Run(fmt.Sprintf("%d bytes", eventSize), func(b *testing.B) {
+			runBench(b, eventSize)
+		})
+	}
+}
+
+func runBench(b *testing.B, eventSize int) {
+	ctx := context.Background()
+	c, close := testPubsubClient(ctx, b, testProjectID)
+	defer close()
+
+	topic, err := c.CreateTopic(ctx, testTopic)
+	if err != nil {
+		b.Fatalf("failed to create topic: %v", err)
+	}
+	sub, err := c.CreateSubscription(ctx, testSub, pubsub.SubscriptionConfig{
+		Topic: topic,
+	})
+	if err != nil {
+		b.Fatalf("failed to create subscription: %v", err)
+	}
+
+	maxMsgs := 2 * int64(runtime.NumCPU())
+	processor := &BenchProcessor{
+		processed: semaphore.NewWeighted(maxMsgs),
+	}
+	h := NewHandler(sub, processor, time.Second)
+	h.Start(ctx, func(err error) {})
+	defer h.Stop()
+	if !h.IsAlive() {
+		b.Error("start handler didn't bring it alive")
+	}
+	testEvent := kgcptesting.NewTestEvent(b, eventSize)
+	msg := new(pubsub.Message)
+	if err := cepubsub.WritePubSubMessage(ctx, binding.ToMessage(testEvent), msg); err != nil {
+		b.Fatal(err)
+	}
+	b.ResetTimer()
+	for i := 0; i < b.N; i++ {
+		processor.processed.Acquire(ctx, 1)
+		topic.Publish(ctx, msg)
+	}
+	processor.processed.Acquire(ctx, maxMsgs)
```

Comment:
IIUC, the semaphore is limiting the maximum number of outstanding messages to `2*cores`, and waiting until the topic is empty before moving on to the next bench run. I'd love to see a quick overview of the theory behind this in a comment. Why `2*cores` as the max concurrency? Is that chosen to normalize resource differences between environments?

Reply:
It wasn't chosen very scientifically, I wanted to choose a number which would buffer enough messages to keep all threads busy while ensuring that the memory footprint used to store published messages does not grow with b.N.